pms/models/pms_room_type.py

```diff
@@ -70,20 +71,91 @@ class PmsRoomType(models.Model):
         "Use `-1` for managing no quota.",
     )
 
-    _sql_constraints = [
-        (
-            "code_type_pms_unique",
-            "unique(code_type)",
-            "Room Type Code must be unique",
-        ),
-    ]
-
     # Constraints and onchanges
     @api.depends("room_ids", "room_ids.active")
     def _compute_total_rooms(self):
         for record in self:
             record.total_rooms_count = len(record.room_ids)
 
+    @api.model
+    def get_unique_by_code_property(self, code_type, pms_property_id):
+        """
+        :param code_type: room type code
+        :param pms_property_id: property ID
+        :return: one or 0 room types ValidationError if more than one found
+        """
+        company_id = self.env["pms.property"].browse(pms_property_id).company_id.id
+        domains = [
+            [
+                ("pms_property_ids", "in", pms_property_id),
+            ],
+            [
+                ("pms_property_ids", "=", False),
+                ("company_id", "=", company_id),
+            ],
+            [
+                ("pms_property_ids", "=", False),
+                ("company_id", "=", False),
+            ],
+        ]
+        for domain in domains:
```

Comment:
Curiosity: Why a for instead of or domain? Does it have a technical reason or is it personal preference?

Reply:
Because we need to search for one domain at a time in order to get the more specific match first.

Imagine we have the following room type in the database:

r1: code_type=c1, properties=[p1], company=False
r2: code_type=c1, properties=False, company=False

and we're searching for (code_type=c1, property=p1)

If we were used an "or" domain, the search would have returned both matches instead of just the first one (the more specific one)

Reply:
Because we need to search for one domain at a time in order to get the more specific match first.

Imagine we have the following room type in the database:

r1: code_type=c1, properties=[p1], company=False
r2: code_type=c1, properties=False, company=False

and we're searching for (code_type=c1, property=p1)

If we were used an "or" domain, the search would have returned both matches instead of just the first one (the more specific one)